File: src/compiler/jdkImage.test.ts
import { test } from "node:test";
import TempDir from "../TempDir.ts";

import { expect } from "expect";

import { execFileSync } from "node:child_process";
import { existsSync, mkdirSync, realpathSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";

import { createChecker } from "./checker.ts";
import { classFilesToStub } from "./classfileReader.ts";
import { emitSourceFile } from "./emitter.ts";
import { createJdkImage } from "./jdkImage.ts";
import { createJdkTypeResolver, installJdkTypes } from "./jdkTypes.ts";
import { createProgram, type Fqn, type PackageName } from "./program.ts";
import { loadJdkStub } from "./jdkStub.ts";
import { type Uri } from "../workspace.ts";
import { writeZip } from "./zipWriter.ts";

// --- hermetic JDK image: synthesize a .jmod from emitted classes, no JDK -------

// Compile Java to .class bytes with our own emitter (no JDK).
function emitClasses(name: string, source: string): { binaryName: string; bytes: Uint8Array }[] {
  const program = createProgram();
  loadJdkStub(program);
  program.setOpenDocument(`file:///${name}.java` as Uri, source, 1);
  const checker = createChecker(program);
  const classes = emitSourceFile(
    program.getSourceFile(`file:///${name}.java` as Uri)!,
    program,
    checker,
  );
  return classes.map(c => ({ binaryName: c.name, bytes: c.bytes }));
}

// A .jmod is the 4-byte magic "JM\x01\x00" followed by a zip whose class entries
// live under classes/ - exactly what jdkImage strips and reads.
function makeJmod(classes: { binaryName: string; bytes: Uint8Array }[]): Uint8Array {
  const zip = writeZip(
    classes.map(c => ({ name: `classes/${c.binaryName}.class`, bytes: c.bytes })),
  );
  const out = new Uint8Array(4 + zip.length);
  out.set([0x4a, 0x4d, 0x01, 0x00], 0);
  out.set(zip, 4);
  return out;
}

// A throwaway JDK home: <tmp>/jmods/<each file>. Auto-removed at process exit.
const tempHomes: string[] = [];
function makeJdkHome(jmods: { name: string; bytes: Uint8Array }[]): string {
  using home = TempDir.create("jdkhome-");
  tempHomes.push(home.path);
  mkdirSync(join(home.path, "jmods"));
  for (const m of jmods) writeFileSync(join(home.path, "jmods", m.name), m.bytes);
  return home.path;
}
process.on("exit", () => {
  for (const home of tempHomes) rmSync(home, { recursive: true, force: true });
});

// A JDK home with jmods/ to read real classes from. Best-effort: JAVA_HOME, then
// javac resolved through PATH. The whole file skips when none is found (CI/Go).
function findJdkHomeWithJmods(): string | undefined {
  const candidates: string[] = [];
  if (process.env.JAVA_HOME) candidates.push(process.env.JAVA_HOME);
  try {
    const finder = process.platform === "win32" ? "where" : "which";
    const javac = execFileSync(finder, ["javac"]).toString().split("\n")[0].trim();
    if (javac) candidates.push(dirname(dirname(realpathSync(javac))));
  } catch {
    // no javac on PATH
  }
  return candidates.find(home => existsSync(join(home, "jmods")));
}

const jdkHome = findJdkHomeWithJmods();
const skip = jdkHome ? false : "no JDK with jmods/ on this machine";

test("the image reads a real JDK class out of its jmods", { skip }, () => {
  const image = createJdkImage(jdkHome!)!;
  expect(image).toBeDefined();

  const family = image.readClassFamily("java/util/List");
  expect(family).toBeDefined();
  const stub = classFilesToStub(family!)!;
  expect(stub.name).toBe("java/util/List");
  expect(stub.source).toContain("package java.util;");
  expect(stub.source).toContain("interface List");

  // A class with a nested type (Map.Entry) folds the nested in from its sibling
  // .class so members resolve.
  const mapFamily = image.readClassFamily("java/util/Map")!;
  expect(mapFamily.length).toBeGreaterThan(1);
  expect(classFilesToStub(mapFamily)!.source).toContain("Entry");

  // A type that does not exist is undefined (no crash, no false positive).
  expect(image.readClassFamily("java/util/NotARealType")).toBeUndefined();
});

test("a consumer resolves JDK types the synthetic stub omits", { skip }, () => {
  const image = createJdkImage(jdkHome!)!;
  const program = createProgram();
  program.setJdkTypeResolver(createJdkTypeResolver(image));
  const index = program.getGlobalIndex();

  // Streams and java.time are absent from jdkStub.ts but present in the image.
  expect(index.getType("java.util.stream.Stream" as Fqn)).toBeDefined();
  expect(index.getType("java.time.LocalDate" as Fqn)).toBeDefined();
  // Common types still resolve.
  expect(index.getType("java.util.List" as Fqn)).toBeDefined();
  expect(index.getType("java.lang.String" as Fqn)).toBeDefined();

  // End to end: a source file referencing a stub-omitted type type-checks with
  // no unresolved-type diagnostic.
  program.setOpenDocument(
    "file:///App.java" as Uri,
    "import java.time.LocalDate;\nclass App { LocalDate today() { return null; } }",
    1,
  );
  const checker = createChecker(program);
  const sourceFile = program.getSourceFile("file:///App.java" as Uri)!;
  const diagnostics = checker.getSemanticDiagnostics(sourceFile);
  const unresolved = diagnostics.filter(d => /resolve|unknown|cannot find/i.test(d.messageText));
  expect(unresolved).toEqual([]);
});

// --- hermetic edge cases (synthetic jmods, no JDK needed) ----------------------

test("createJdkImage returns undefined when there is nothing to read", () => {
  // No jmods/ directory at all.
  using noJmods = TempDir.create("nojmods-");
  tempHomes.push(noJmods.path);
  expect(createJdkImage(noJmods.path)).toBeUndefined();
  // A jmods/ directory with no .jmod files in it.
  expect(createJdkImage(makeJdkHome([]))).toBeUndefined();
});

test("a non-jmod file in jmods/ is skipped, not crashed on", () => {
  // The file ends in .jmod (so createJdkImage sees a candidate) but lacks the
  // JM magic, so reading it yields nothing rather than throwing.
  const home = makeJdkHome([{ name: "junk.jmod", bytes: new Uint8Array([1, 2, 3, 4, 5]) }]);
  const image = createJdkImage(home)!;
  expect(image).toBeDefined();
  expect(image.readClassFamily("lib/Whatever")).toBeUndefined();
});

test("readClassFamily resolves a class, misses cleanly, and handles the default package", () => {
  const home = makeJdkHome([
    makeJmodEntry(
      "lib/Widget",
      "package lib;\npublic class Widget { public int size() { return 0; } }",
    ),
    makeJmodEntry("Root", "public class Root { public int v() { return 0; } }"), // default package
  ]);
  const image = createJdkImage(home)!;

  const widget = image.readClassFamily("lib/Widget");
  expect(widget).toBeDefined();
  expect(classFilesToStub(widget!)!.name).toBe("lib/Widget");

  // Default-package class (no slash in the binary name).
  expect(classFilesToStub(image.readClassFamily("Root")!)!.name).toBe("Root");

  // A class that is not present.
  expect(image.readClassFamily("lib/Missing")).toBeUndefined();
});

test("a nested class folds into its outer's family", () => {
  const classes = emitClasses(
    "Outer",
    [
      "package lib;",
      "public class Outer {",
      "  public static class Builder { public int knobs; }",
      "}",
    ].join("\n"),
  );
  // Sanity: the emitter produced both the outer and the nested class.
  expect(classes.map(c => c.binaryName).sort()).toEqual(["lib/Outer", "lib/Outer$Builder"]);

  const image = createJdkImage(makeJdkHome([{ name: "lib.jmod", bytes: makeJmod(classes) }]))!;
  const family = image.readClassFamily("lib/Outer")!;
  expect(family.length).toBe(2); // outer + nested
  expect(classFilesToStub(family)!.source).toContain("class Builder");
});

test("a project type shadows a JDK type of the same name", () => {
  // The image carries lib.Thing with imageOnly(); the project declares its own
  // lib.Thing with projectOnly(). getType must return the project's.
  const image = createJdkImage(
    makeJdkHome([
      makeJmodEntry(
        "lib/Thing",
        "package lib;\npublic class Thing { public int imageOnly() { return 0; } }",
      ),
    ]),
  )!;
  const program = createProgram();
  program.setJdkTypeResolver(createJdkTypeResolver(image));
  program.addProjectFile(
    "file:///lib/Thing.java" as Uri,
    "package lib;\npublic class Thing { public int projectOnly() { return 1; } }",
  );
  const thing = program.getGlobalIndex().getType("lib.Thing" as Fqn);
  expect(thing?.members?.has("projectOnly")).toBe(true);
  expect(thing?.members?.has("imageOnly")).toBe(false);
});

test("a JDK-type miss is idempotent (cached, no re-read)", () => {
  const home = makeJdkHome([makeJmodEntry("lib/Only", "package lib;\npublic class Only {}")]);
  const resolve = createJdkTypeResolver(createJdkImage(home)!);
  expect(resolve("lib.Absent" as Fqn)).toBeUndefined();
  expect(resolve("lib.Absent" as Fqn)).toBeUndefined(); // second call hits the null cache
  expect(resolve("lib.Only" as Fqn)).toBeDefined();
});

test("installJdkTypes falls back to the synthetic stub with no provisioned JDK", () => {
  // No config (the LSP can run without one): the stub must still resolve.
  const program = createProgram();
  installJdkTypes(program, undefined);
  expect(program.getGlobalIndex().getType("java.lang.String" as Fqn)).toBeDefined();
});

test("a provisioned JDK keeps the stub for enumeration and adds image resolution", () => {
  // The real install path loads the stub (so completion / auto-import keep
  // enumerating) AND sets the image resolver (so stub-omitted types resolve).
  // Mirror that composition with a synthetic image carrying an extra type.
  const home = makeJdkHome([makeJmodEntry("lib/Extra", "package lib;\npublic class Extra {}")]);
  const program = createProgram();
  loadJdkStub(program);
  program.setJdkTypeResolver(createJdkTypeResolver(createJdkImage(home)!));
  const index = program.getGlobalIndex();

  // Enumeration (completion / auto-import) still sees the stub's common types.
  expect(index.getPackageTypes("java.lang" as PackageName)?.size ?? 0).toBeGreaterThan(0);
  expect(index.findFqnsBySimpleName("String").length).toBeGreaterThan(0);
  // Resolution reaches the image for types the stub does not carry.
  expect(index.getType("lib.Extra" as Fqn)).toBeDefined();
});

// Build one synthetic jmod holding a single emitted class, named after it.
function makeJmodEntry(binaryName: string, source: string): { name: string; bytes: Uint8Array } {
  const simple = binaryName.slice(binaryName.lastIndexOf("/") + 1);
  return { name: `${simple}.jmod`, bytes: makeJmod(emitClasses(simple, source)) };
}

File: togo/internal/compiler/jdk_image_test.go
package compiler

// Port of src/compiler/jdkImage.test.ts. Reads real classes out of a JDK's
// jmods/; the whole file skips when no JDK with jmods/ is found (so CI without a
// JDK still passes).

import (
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
)

// --- hermetic JDK image: synthesize a .jmod from emitted classes, no JDK -------

// emitClasses compiles Java to .class bytes with our own emitter (no JDK).
func emitClasses(t *testing.T, name, source string) []EmittedClass {
	t.Helper()
	program := NewProgram()
	LoadJdkStub(program)
	uri := URI("file:///" + name + ".java")
	program.SetOpenDocument(uri, source, 1)
	checker := NewChecker(program)
	return EmitSourceFile(program.GetSourceFile(uri), program, checker, false)
}

// makeJmod is the 4-byte magic "JM\x01\x00" followed by a zip whose class entries
// live under classes/ - exactly what jdk_image strips and reads.
func makeJmod(classes []EmittedClass) []byte {
	entries := make([]ZipEntryInput, 0, len(classes))
	for _, c := range classes {
		entries = append(entries, ZipEntryInput{Name: "classes/" + c.Name + ".class", Bytes: c.Bytes})
	}
	return append([]byte{0x4A, 0x4D, 0x01, 0x00}, WriteZip(entries)...)
}

// makeJdkHome writes a throwaway JDK home <tmp>/jmods/<each file>.
func makeJdkHome(t *testing.T, jmods map[string][]byte) string {
	t.Helper()
	home := t.TempDir()
	if err := os.Mkdir(filepath.Join(home, "jmods"), 0o755); err != nil {
		t.Fatal(err)
	}
	for name, bytes := range jmods {
		if err := os.WriteFile(filepath.Join(home, "jmods", name), bytes, 0o644); err != nil {
			t.Fatal(err)
		}
	}
	return home
}

// makeJmodEntry builds one synthetic jmod holding a single emitted class.
func makeJmodEntry(t *testing.T, binaryName, source string) (string, []byte) {
	t.Helper()
	simple := binaryName[strings.LastIndexByte(binaryName, '/')+1:]
	return simple + ".jmod", makeJmod(emitClasses(t, simple, source))
}

// findJdkHomeWithJmods returns a JDK home containing jmods/, best-effort:
// JAVA_HOME, then javac resolved through PATH. Empty when none is found.
func findJdkHomeWithJmods() string {
	var candidates []string
	if h := os.Getenv("JAVA_HOME"); h != "" {
		candidates = append(candidates, h)
	}
	if javac, err := exec.LookPath("javac"); err == nil {
		if real, err := filepath.EvalSymlinks(javac); err == nil {
			candidates = append(candidates, filepath.Dir(filepath.Dir(real)))
		}
	}
	for _, home := range candidates {
		if _, err := os.Stat(filepath.Join(home, "jmods")); err == nil {
			return home
		}
	}
	return ""
}

func TestJdkImageReadsRealClass(t *testing.T) {
	home := findJdkHomeWithJmods()
	if home == "" {
		t.Skip("no JDK with jmods/ on this machine")
	}
	image := NewJdkImage(home)
	if image == nil {
		t.Fatal("NewJdkImage returned nil for a JDK with jmods/")
	}

	family := image.ReadClassFamily("java/util/List")
	if family == nil {
		t.Fatal("java/util/List not found in image")
	}
	stub, ok := ClassFilesToStub(family)
	if !ok {
		t.Fatal("ClassFilesToStub failed for java/util/List")
	}
	if stub.Name != "java/util/List" {
		t.Errorf("stub name = %q, want java/util/List", stub.Name)
	}
	if !strings.Contains(stub.Source, "package java.util;") || !strings.Contains(stub.Source, "interface List") {
		t.Errorf("unexpected stub source:\n%s", stub.Source)
	}

	// Map has a nested type (Map.Entry) in a sibling .class; the family folds it in.
	mapFamily := image.ReadClassFamily("java/util/Map")
	if len(mapFamily) <= 1 {
		t.Errorf("java/util/Map family = %d classes, want > 1 (nested Entry)", len(mapFamily))
	}
	mapStub, _ := ClassFilesToStub(mapFamily)
	if !strings.Contains(mapStub.Source, "Entry") {
		t.Error("Map stub should contain nested Entry")
	}

	// A type that does not exist is nil (no crash, no false positive).
	if image.ReadClassFamily("java/util/NotARealType") != nil {
		t.Error("a missing type should read back as nil")
	}
}

func TestConsumerResolvesStubOmittedJdkTypes(t *testing.T) {
	home := findJdkHomeWithJmods()
	if home == "" {
		t.Skip("no JDK with jmods/ on this machine")
	}
	image := NewJdkImage(home)
	program := NewProgram()
	program.SetJdkTypeResolver(createJdkTypeResolver(image))
	index := program.GetGlobalIndex()

	// Streams and java.time are absent from jdkstub.go but present in the image.
	for _, fqn := range []Fqn{
		"java.util.stream.Stream", "java.time.LocalDate", "java.util.List", "java.lang.String",
	} {
		if index.GetType(fqn) == nil {
			t.Errorf("GetType(%q) = nil, want resolved from the image", fqn)
		}
	}

	// End to end: a source referencing a stub-omitted type type-checks with no
	// unresolved-type diagnostic.
	program.SetOpenDocument("file:///App.java",
		"import java.time.LocalDate;\nclass App { LocalDate today() { return null; } }", 1)
	checker := NewChecker(program)
	for _, d := range checker.GetSemanticDiagnostics(program.GetSourceFile("file:///App.java")) {
		if strings.Contains(strings.ToLower(d.MessageText), "resolve") ||
			strings.Contains(strings.ToLower(d.MessageText), "cannot find") {
			t.Errorf("unexpected unresolved diagnostic: %s", d.MessageText)
		}
	}
}

// --- hermetic edge cases (synthetic jmods, no JDK needed) ----------------------

func TestNewJdkImageNilWhenNothingToRead(t *testing.T) {
	// No jmods/ directory at all.
	if NewJdkImage(t.TempDir()) != nil {
		t.Error("a home without jmods/ should give a nil image")
	}
	// A jmods/ directory with no .jmod files in it.
	if NewJdkImage(makeJdkHome(t, nil)) != nil {
		t.Error("an empty jmods/ should give a nil image")
	}
}

func TestNonJmodFileIsSkipped(t *testing.T) {
	// The file ends in .jmod (so NewJdkImage sees a candidate) but lacks the JM
	// magic, so reading it yields nothing rather than panicking.
	home := makeJdkHome(t, map[string][]byte{"junk.jmod": {1, 2, 3, 4, 5}})
	image := NewJdkImage(home)
	if image == nil {
		t.Fatal("a candidate .jmod file should still yield an image")
	}
	if image.ReadClassFamily("lib/Whatever") != nil {
		t.Error("a junk jmod should read back as nil")
	}
}

func TestReadClassFamilyResolvesMissesAndDefaultPackage(t *testing.T) {
	wName, wBytes := makeJmodEntry(t, "lib/Widget", "package lib;\npublic class Widget { public int size() { return 0; } }")
	rName, rBytes := makeJmodEntry(t, "Root", "public class Root { public int v() { return 0; } }")
	image := NewJdkImage(makeJdkHome(t, map[string][]byte{wName: wBytes, rName: rBytes}))

	widget := image.ReadClassFamily("lib/Widget")
	if widget == nil {
		t.Fatal("lib/Widget should resolve")
	}
	if stub, ok := ClassFilesToStub(widget); !ok || stub.Name != "lib/Widget" {
		t.Errorf("stub = %q (ok=%v), want lib/Widget", stub.Name, ok)
	}
	// Default-package class (no slash in the binary name).
	if stub, ok := ClassFilesToStub(image.ReadClassFamily("Root")); !ok || stub.Name != "Root" {
		t.Errorf("default-package stub = %q (ok=%v), want Root", stub.Name, ok)
	}
	// A class that is not present.
	if image.ReadClassFamily("lib/Missing") != nil {
		t.Error("a missing class should read back as nil")
	}
}

func TestNestedClassFoldsIntoOuterFamily(t *testing.T) {
	classes := emitClasses(t, "Outer", strings.Join([]string{
		"package lib;",
		"public class Outer {",
		"  public static class Builder { public int knobs; }",
		"}",
	}, "\n"))
	if len(classes) != 2 {
		t.Fatalf("emitter produced %d classes, want 2 (outer + nested)", len(classes))
	}
	image := NewJdkImage(makeJdkHome(t, map[string][]byte{"lib.jmod": makeJmod(classes)}))
	family := image.ReadClassFamily("lib/Outer")
	if len(family) != 2 {
		t.Fatalf("family = %d classes, want 2", len(family))
	}
	stub, _ := ClassFilesToStub(family)
	if !strings.Contains(stub.Source, "class Builder") {
		t.Error("outer stub should fold in the nested Builder")
	}
}

func TestProjectTypeShadowsJdkType(t *testing.T) {
	// The image carries lib.Thing with imageOnly(); the project declares its own
	// lib.Thing with projectOnly(). GetType must return the project's.
	name, bytes := makeJmodEntry(t, "lib/Thing", "package lib;\npublic class Thing { public int imageOnly() { return 0; } }")
	image := NewJdkImage(makeJdkHome(t, map[string][]byte{name: bytes}))
	program := NewProgram()
	program.SetJdkTypeResolver(createJdkTypeResolver(image))
	program.AddProjectFile("file:///lib/Thing.java",
		"package lib;\npublic class Thing { public int projectOnly() { return 1; } }")
	thing := program.GetGlobalIndex().GetType("lib.Thing")
	if thing == nil || thing.Members["projectOnly"] == nil {
		t.Fatal("GetType should return the project's lib.Thing")
	}
	if thing.Members["imageOnly"] != nil {
		t.Error("the image's lib.Thing must not shadow the project's")
	}
}

func TestJdkTypeMissIsIdempotent(t *testing.T) {
	name, bytes := makeJmodEntry(t, "lib/Only", "package lib;\npublic class Only {}")
	resolve := createJdkTypeResolver(NewJdkImage(makeJdkHome(t, map[string][]byte{name: bytes})))
	if resolve("lib.Absent") != nil {
		t.Error("absent type should be nil")
	}
	if resolve("lib.Absent") != nil {
		t.Error("absent type should stay nil on the cached second call")
	}
	if resolve("lib.Only") == nil {
		t.Error("present type should resolve")
	}
}

func TestInstallJdkTypesFallsBackToStub(t *testing.T) {
	// No config (the LSP can run without one): the stub must still resolve.
	program := NewProgram()
	InstallJdkTypes(program, nil)
	if program.GetGlobalIndex().GetType("java.lang.String") == nil {
		t.Error("with no JDK, java.lang.String should resolve from the stub")
	}
}
